### DISCUSSION TITLE
SignalR no connection in `ws` mode

### DISCUSSION THREAD

**Discussion by mtlive** [ThreeMammals/Ocelot#2096] [Show and tell]
I tried everything in documentation for using [SignalR](https://ocelot.readthedocs.io/en/latest/features/websockets.html#signalr) in websocket mode (`ws` scheme) but couldn't get it to work.
In our staging servers I get the "Handshake was canceled error" and in debug it takes so long that it switches out from ws.

## Actual Behavior 
![image](https://github.com/ThreeMammals/Ocelot/assets/2260557/fd05696a-03bc-47ae-933c-2376a0c9e10a)

## Specifications

  - Version: 23.3.0
  - Platform: .NET 8 (Windows)
  - Subsystem:

Also added   services.AddSignalR();
Confiuration:
```json
{
"Routes": [
   {
     "DownstreamHostAndPorts": [
       {
         "Host": "192.168.10.209",
         "Port": 5003
       }
     ],
     "DownstreamPathTemplate": "/{everything}",
     "DownstreamScheme": "ws",
     "UpstreamPathTemplate": "/chat/{everything}",
   },
],
"DynamicRoutes": [],
"Aggregates": [],
"GlobalConfiguration": {
  "RequestIdKey": null,
  "ServiceDiscoveryProvider": {
    "Scheme": null,
    "Host": null,
    "Port": 0,
    "Type": null,
    "Token": null,
    "ConfigurationKey": null,
    "PollingInterval": 0,
    "Namespace": null
  },
  }
}
```

**Comment by raman-m** [comment]
Hi @mtlive !

> I tried everything in documentation for using [SignalR](https://ocelot.readthedocs.io/en/latest/features/websockets.html#signalr) in websocket mode but couldn't get it to work.

Please note that we are not a consultancy team. I trust the community will assist you.

> In our staging servers I get the "Handshake was canceled error" and in debug it takes so long that it switches out from `ws`.

Unfortunately, our repository lacks a robust [sample](https://github.com/ThreeMammals/Ocelot/tree/develop/samples), and the [Websockets](https://ocelot.readthedocs.io/en/latest/features/websockets.html#websockets) feature for [SignalR](https://ocelot.readthedocs.io/en/latest/features/websockets.html#signalr) is a complex functionality to implement. It would be more efficient if you could upload your solution to your account as a repository for us to review.

**Reply by raman-m** [reply]
Regarding your settings JSON...
Why did you specify `"ServiceDiscoveryProvider"` section in global configuration?

Remove `"DynamicRoutes"` and `"Aggregates"` as well.
But the route definition looks good.

What's your `"BaseUrl"` in global configuration?
Did you setup Ocelot endpoint as WS one?

**Comment by raman-m** [comment]
I recommend conducting a Bing [search for ".NET websocket 'Handshake was canceled error'"](https://www.bing.com/search?q=.NET+websocket+%22Handshake+was+canceled+error%22) now.
Unfortunately, I don't have time to spend on a solution that I can't see.

**Comment by mtlive** [comment]
The issue was the wrong order of middleware.

**Reply by raman-m** [reply]
Congrats! 🎉
